WIP: Implement RV32C instruction decoder

ccs100203 opened the pull request:
# RISC-V Compressed Instruction Support

We implemented the compressed instructions for rv32emu-next.
Moreover, We still don't support RV32C.F instructions.
However, it doesn't yet pass the compliance test.

Commit:
Modified .gitignore for testing files

ccs100203 committed:
Revise PC, Add Inst. Table & Inst. Distinguishment

In `riscv_private.h`:
- Add variable `inst_len` in structure `riscv_t` to record length of current instruction
- Add mask for compressed instruction opcode

In `riscv.c`:
- Add function table `c_opcodes`, defined `NULL`
- Modify `rv_step()` to process compressed instructions

ccs100203 committed:
Add Decoding Func., Inst. Templates & A Few Func. Definitions

In riscv_private.h:
- Add instruction decoding function & revise masks

In io.c:
- revise address check

In riscv.c:
- Add compressed instruction templates
- Add register checking function
- Implement c_op_addi(), c_op_swsp(), c_op_addi4spn(), c_op_li()

Commit:
Modified Makefile to support debug build & RV32C

In `Makefike`:
- Added target `debug`, used to enable debug functions when building
- Added `ENABLE_RV32C` flag

Commit:
Implemented c_op_lwsp(), Wrapped RV32F Instructions.

In `riscv.c`:
- Implemented `c_op_lwsp()` for c.lwsp instruction
- Moved F-Extension related functions into #ifdef wrappers for future implementation

Commit:
Implemented `c_op_lw()`, Added debug messages

In `riscv.c`:
- Added debug message "Entered c.XXX" to inform which function is called
- Implemented `c_op_lw()`
- Removed RV64 and RV128 only instructions
- Moved more RV32F related instruction into #define wrapper
- Added temporary variable to some function to reduce line width

Commit:
Inplemented CJ-type instructions.

In `riscv.c`:
- Inplemented `c_op_j()` , `c_op_jal()`

Commit:
Modified target degug to make sure rebuild

Commit:
Implemented CJ-type inst.

In `riscv.c`
- Added `c_op_cjtype()`
- Added instruction unaligned exception check for PC modifying instructions

In `riscv_private.h`:
- Removed unneeded masks

TODO:
- Find out rules for instruction functions' (op_XXX) return value

ccs100203 committed:
Revise Debug Function

In riscv.c:
- Modify print_register() to being a macro

ccs100203 committed:
Add Define NULL in Compressed Inst.

In risc.c:
- Add compressed instruction define NULL

ccs100203 committed:
Revise Compressed Float Inst. Definition

In riscv.c:
- Extract RV32C.F from RV32C block

ccs100203 committed:
Implement C.LUI & C.ADDI16SP Inst.

In riscv.c:
- Implement c_op_lui function
- Revise a few sign-extended immediates

Commit:
Implemented CB-type Inst, Fixed Few Sign-Extension Bugs

in `riscv.c`:
- Implemented c_bneq(), c_beqz()
- Fixed overlooked sign extension bugs

in `riscv-private.c`:
- Added c_dec_cbtype_imm(), c_dec_cjtype_imm()

Commit:
Implemented c.sw Instruction

In `riscv.c`:
- Added c_op_sw();

Commit:
Implemented c_op_slli()

in `riscv.c`:
- Added c_op_slli()

Commit:
Implement c.srli, c.srai, c.slli c.andi Inst.

In `riscv.c`:
- Started implementing c_op_misc_alu()
- Added c_op_slli()
- Added c_op_srli()
- Added c_op_srai()
- Added c_op_andi()

TODO:
- Change name of c_op_cr to fit future modification (add C.MV and C.ADD, etc.)

Commit:
Relocated some debug print

Commit:
Implemented c.sub, c.and, c.xor, c.or Inst.

Commit:
Implemented c.mv, c.add Inst.

In `riscv.c`:
- Revised c_op_cr() to fit all 5 instructions:
  + c.add
  + c.mv
  + c.jr
  + c.jalr
  + c.ebreak (Debug message only)
- Changed return value to false wherever PC might jump

ccs100203 committed:
Fix Inst. Ebreak Function

In riscv.c:
- Add the ebreak function

ccs100203 committed:
Remove Unuseful Comment

In riscv.c:
- remove unuseful comment

jserv commented on .gitignore:
```diff
@@ -2,3 +2,7 @@ build/DOOM1.WAD*
 *.o
 *.o.d
+
+# Added by Uduru0522
```
You don't have to mention your name since git would track.

jserv commented on Makefile:
```diff
@@ -68,4 +69,8 @@ clean:
 distclean: clean
 	$(RM) $(OUT)/DOOM1.WAD $(OUT)/DOOM1.WAD.sha1
 
+debug: clean debug_setting 
```
This change should only appear in another pull request.

ccs100203 replied:
Removed

jserv commented on io.c:
```diff
@@ -70,7 +70,7 @@ uint32_t memory_read_str(memory_t *m, uint8_t *dst, uint32_t addr, uint32_t max)
 {
     const uint32_t addr_lo = addr & mask_lo;
-    assert((addr_lo & 3) == 0);
+    assert((addr_lo & 1) == 0);
```
Why?

ccs100203 replied:
After executing a compressed instruction, the Program Counter might plus 2, and we can't promise that the last two bits of address always remain 0.  Thus, we modified `(addr_lo & 3)` into `(addr_lo & 1)`.

jserv commented on riscv.c:
```diff
@@ -770,8 +791,485 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+
+static bool c_op_addi4spn(struct riscv_t *rv, uint16_t inst)
+{
+    debug_print("Entered c.addi4spn");
```
You can use GDB script to locate the instruction handlers instead of dumping the craft.

jserv commented on riscv.c:
```diff
@@ -770,8 +791,485 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+
+    switch ((inst & 0x1000) >> 12) {
+    case 0:  
+        if(rs2){
```
Use `clang-format -i` for consistent style.

ccs100203 committed:
Revise Coding Format

Using clang-format to format the code.

ccs100203 committed:
Remove Useless Comment

Remove useless comment in .gitignore.

ccs100203 committed:
Revise Address Validation

Adding a condition, if RV is enabled, the addr_io check will be changed into 1.

ccs100203 committed:
Remove Debug Setting & Transfer to Other Branch

The debug setting and dumping is transferred to 'debug' branch.

jserv commented on .gitignore:
```diff
@@ -2,3 +2,5 @@ build/DOOM1.WAD*
 build/rv32emu
 *.o
 *.o.d
+test/
```
Rename `test` to directory `tests`.

jserv commented on riscv_private.h:
```diff
@@ -210,3 +238,67 @@ static inline uint32_t sign_extend_b(uint32_t x)
+    temp |= (x & 0b0000000001100000) << 1;
+    temp |= (x & 0b0001000000000000) >> 4;
+    // extend to 16 bit
+    for(int i = 1; i < 8; ++i){
```
Use `clang-format -i` to ensure the consistent styie.

ccs100203 replied:
It's my negligence, I'm sorry.

jserv commented on riscv.c:
```diff
@@ -770,8 +771,440 @@ static bool op_amo(struct riscv_t *rv, uint32_t inst)
+    const uint32_t imm = sign_extend_h(c_dec_cbtype_imm(inst));
+    const uint32_t rs1 = c_dec_rs1c(inst) | 0x08;
+
+    if (rv->X[rs1]) {
```
Rewrite as following:
```c
rv->PC += rv->X[rs1] ? imm : rv->inst_len;
```

ccs100203 committed:
Revise Code Format

Reactivate clang in riscv_private.h, make it can be formated by clang.

ccs100203 committed:
Rename a Directory in .gitignore

Rename test/ to tests/.

ccs100203 committed:
Compact the If Condition & Add Some Comments

Rewrite c_op_beqz & c_op_bnez into a ternary operator.

jserv commented:
Drop in favor of #11
